test: tool wallet test coverage for unexpected writes to wallet

jonatack opened the pull request:
This pull request adds test coverage in `test/functional/tool_wallet.py` to reproduce unexpected writes to the wallet as described in https://github.com/bitcoin/bitcoin/issues/15608 and serve as a benchmark for fixing the issue:

- Wallet tool `info` unexpectedly writes to the wallet file if the wallet file permissions are read/write.

- Wallet tool `info` raises with "Error loading . Is wallet being used by another process?" if the wallet file permissions are read-only.

Goals:

1. Reproduce the reported issue, define the current unexpected behavior, and add test coverage to guide a future fix. Add debug-level logging for sanity checking and commented-out assertions to be uncommented when fixing the issue. Add the same coverage to the wallet tool create test and the getwalletinfo test as regression tests while fixing the issue.

2. Add info log messages as there are currently none in the test file.

3. Split the tests out to separate functions as per review feedback.

Thanks to Marco Falke for pointing me in the right direction.

jonatack committed:
test: Add log messages to test/functional/tool_wallet.py

and update code comments as per Python PEP 8 style guide.

jonatack commented on test/functional/tool_wallet.py:
```diff
@@ -36,8 +39,29 @@ def assert_tool_output(self, output, *args):
 
-    def run_test(self):
+    def wallet_path(self):
+        return os.path.join(self.nodes[0].datadir, "regtest", "wallets", "wallet.dat")
```
Use single quotes here like the rest of the file.

MarcoFalke commented on test/functional/tool_wallet.py:
```diff
@@ -36,8 +39,29 @@ def assert_tool_output(self, output, *args):
 
+    def log_wallet_timestamp_comparison(self, new, old):
+        result = 'unchanged' if new == old else 'increased! **FIX ME**'
+        self.log.info('Wallet file timestamp {0}\n'.format(result))
```
nit: instead of logging the timestamps and whether they changed, you might use `assert_equal` (or explicitly comment it out, if it currently fails)

jonatack replied:
Thanks for reviewing! I added both logs and assertions with the logs serving as a sanity check for the assertions. Based on your feedback, I switched the assertions from passing ones that need to be inverted, to failing ones that need to be uncommented.

MarcoFalke commented on test/functional/tool_wallet.py:
```diff
@@ -46,9 +70,23 @@ def run_test(self):
+        os.chmod(self.wallet_path(), 0o400)
+        self.log.info('Call wallet tool info')
+        self.assert_raises_tool_error('Error loading . Is wallet being used by another process?', 'info')
+        self.log.info('Wallet tool info command raises loading error! **FIX ME**')
```
TODO's shouldn't be logged, but added as inline comments

jonatack replied:
Done. Waiting for Travis CI to pass locally and then will push a rebased update.

jonatack committed:
test: Tool wallet test coverage for unexpected writes to wallet

This commit adds test coverage in `test/functional/tool_wallet.py` to reproduce unexpected writes to the wallet as described in https://github.com/bitcoin/bitcoin/issues/15608:

    - wallet tool `info` unexpectedly writes to the wallet file if the wallet file permissions are read/write.

    - wallet tool `info` raises with "Error loading . Is wallet being used by another process?" if the wallet file permissions are read-only.

1. Reproduce the reported issue, define the current unexpected behavior, and add test coverage to guide a future fix in the form of commented-out assertions to be uncommented when testing/fixing.

2. Provisionally extend the same coverage to the wallet tool create test and the getwalletinfo test as regression tests while fixing the issue.

3. Add some logging for sanity checking.

------

Changes after rebase:

5. Make wallet_path an instance method instead of a function in tool_wallet.py as per Marco Falke review suggestion.

6. Assert wallet permissions instead of logging them in tool_wallet.py. This ran into an issue with Appveyor keeping permissions at 666 so allowed for 666 as a workaround.

7. Change the added logging from info to debug level.

8. More helpful assertions order in tool_wallet.py#assert_tool_output. This change makes #assert_tool_output raise "Error loading wallet.dat. Is wallet being used by another process?" rather than a less-helpful message when debugging the read-only wallet permissions issue.

jonatack commented:
Lowered the buffer size for the shasum function to 16Kb and hoisted that value to a constant. Rebased.

This should be ready for review.

laanwj commented on test/functional/tool_wallet.py:
```diff
@@ -87,15 +161,32 @@ def run_test(self):
         ''')
         self.assert_tool_output(out, '-wallet=foo', 'create')
-
+        shasum_after = self.wallet_shasum()
```
I understand surrounding the `info` methods, but what is the logic here? creating a wallet that already exists should leave it untouched?

jonatack replied:
Yes, updated test name and logging to clarify that a bit. Happy to remove if you reckon it isn't a useful safeguard while fixing the issue.

laanwj commented:
Looks good to me
ACK 96e69a2ace8b0db496a0c429f05ccd5cb2f39c94

MarcoFalke commented on test/functional/tool_wallet.py:
```diff
@@ -36,8 +42,32 @@ def assert_tool_output(self, output, *args):
+
+    def wallet_timestamp(self, op):
+        timestamp = os.path.getmtime(self.wallet_path())
+        self.log.info('Wallet file timestamp {0}: {1}'.format(op, timestamp))
```
style-nit: Is it required to log the timestamp in a simple getter method? I'd prefer if it was logged by the caller or just asserted (Looks like you already do both of that)

jonatack replied:
Done.

MarcoFalke commented on test/functional/tool_wallet.py:
```diff
@@ -36,8 +42,32 @@ def assert_tool_output(self, output, *args):
 
+    def log_wallet_permissions(self):
+        result = oct(os.lstat(self.wallet_path()).st_mode)[-3:]
+        self.log.info('Verifying wallet file permissions: {}'.format(result))
```
You aren't verifying anything here, it seems. I'd prefer if this was just a simple getter util method and the caller logs/asserts if appropriate.

jonatack replied:
Done.

MarcoFalke commented on test/functional/tool_wallet.py:
```diff
@@ -36,8 +42,32 @@ def assert_tool_output(self, output, *args):
+
+    def log_wallet_timestamp_comparison(self, old, new):
+        result = 'unchanged' if new == old else 'increased!'
+        self.log.info('Wallet file timestamp {}\n'.format(result))
```
This looks like a test-dev debug statement. Maybe use `self.log.debug`?

jonatack replied:
Done, switched to loglevel `debug` for all the timestamp and shasum logging.

MarcoFalke commented on test/functional/tool_wallet.py:
```diff
@@ -59,12 +101,32 @@ def run_test(self):
         self.stop_node(0)
 
+        self.log.info('Calling wallet tool info after generating a transaction')
+        shasum_before = self.wallet_shasum()
```
Looks like you are overwriting a previous symbol with the same name. I think it is clearer to split the tests into smaller ones, so that overwriting does not happen. I.e.

```py
  self run_test(self):
   self.run_test_foo_bar()
   self.run_test_gen_tx()

  def run_test_foo_bar(self):
    assert True

...

jonatack replied:
Done.

maflcko reviewed (commented):
some style-nits

jonatack committed:
test: Split tool_wallet.py test into subtests

as per Marco Falke review suggestion.

jonatack commented on test/functional/tool_wallet.py:
```diff
@@ -36,8 +42,32 @@ def assert_tool_output(self, output, *args):
 
-    def run_test(self):
+    def wallet_path(self):
+        return os.path.join(self.nodes[0].datadir, 'regtest', 'wallets', 'wallet.dat')
```
Done.

jonatack commented:
Linter is complaining about unused code because it is in the commented-out tests. Reworking.

jonatack commented:
Appveyor appears to keep file permissions at 666 so allowed for 666 as a workaround.

jonatack commented:
Thanks for the review @laanwj and @MarcoFalke. I've pushed rebased commits that address your feedback. Are you able to rereview/retest?

laanwj commented:
code review ACK 7195fa792fcc19e9c064c4e38814c3b46a210b34

jonatack commented:
The lone Travis failure appears to be an unrelated `wait_until` timeout > 60 sec in `feature_block.py`.
